### PR TITLE
drivers: dma: stm32 initialize a dma_stm32_data structure

### PR DESCRIPTION
**Pull request by FRASTM**
In the dma.h there is a dma_ctx structure using a magic code
to be identify. This structure must be prepared as a new element of the dma_stm32_data.

Without this init, the `tests/drivers/dma/loop_transfer/` on the  test _nucleo_f091rc_ board fails in the `dma_request_channel()
` function call.

```
*** Booting Zephyr OS build zephyr-v2.6.0-1722-g8566bfec2609  *
Running test suite dma_m2m_loop_test                           
===============================================================
START - test_dma_m2m_loop                                      
DMA memory to memory transfer started on DMA_1                 
Preparing DMA Controller                                       
E: ***** HARD FAULT *****                                      
E: r0/a1:  0x080059d4  r1/a2:  0xfffffc7f  r2/a3:  0x20000488  
E: r3/a4:  0x200006b1 r12/ip:  0x00000000 r14/lr:  0x08002d6f  
E:  xpsr:  0x21000000                                          
E: Faulting instruction address (r15/pc): 0x08000550           
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0            
E: Current thread: 0x20000290 (test_dma_m2m_loop)              
E: Halting system      
```

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

**Comment by FRASTM** [comment]
When initialised, the tests/drivers/dma/loop_transfer PASSED on the nucleo_f091rc.
```
*** Booting Zephyr OS build zephyr-v2.6.0-1722-g8566bfec2609  *
Running test suite dma_m2m_loop_test                           
===============================================================
START - test_dma_m2m_loop                                      
DMA memory to memory transfer started on DMA_1                 
Preparing DMA Controller                                       
this platform do not support the dma channel                   
Starting the transfer on channel 1 and waiting for 1 second    
Each RX buffer should contain the full TX buffer string.       
RX data Loop 0: The quick brown fox jumps over the lazy dog    
RX data Loop 1: The quick brown fox jumps over the lazy dog    
RX data Loop 2: The quick brown fox jumps over the lazy dog    
RX data Loop 3: The quick brown fox jumps over the lazy dog    
RX data Loop 4: The quick brown fox jumps over the lazy dog    
Finished: DMA                                                  
 PASS - test_dma_m2m_loop in 1.50 seconds                      
===============================================================
Test suite dma_m2m_loop_test succeeded      
```

**Comment by FRASTM** [comment]
In the dma.h, the dma_ctx->magic is defined as an **int**, and compared to  the 32-bit DMA_MAGIC.
The dma_stm32_data which is mapped to a dma_context is now initialised and test does not crash anymore
```
	if (dma_ctx->magic != DMA_MAGIC) {
		return channel;
	}
```
could

**Review by ABOSTM — APPROVED** [review]
LGTM

File: drivers/dma/dma_stm32.c [hunk 1]
```diff
@@ -610,6 +610,10 @@ static int dma_stm32_init(const struct device *dev)
 #endif /* CONFIG_DMAMUX_STM32 */
 	}
 
+	((struct dma_stm32_data *)dev->data)->dma_ctx.magic = 0;
+	((struct dma_stm32_data *)dev->data)->dma_ctx.dma_channels = 0;
+	((struct dma_stm32_data *)dev->data)->dma_ctx.atomic = 0;
+
```
Comment (erwango) [inline drivers/dma/dma_stm32.c:616]: Can it be moved to `dma_stm32_data_##index` initialization ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_stm32.c (hunk 1 above)
Comment (FRASTM) [reply drivers/dma/dma_stm32.c:616]: put this init in the `static struct dma_stm32_data dma_stm32_data_##index` section

**Comment by FRASTM** [comment]
move the initialisation  of the `dma_stm32_data` in the static struct dma_stm32_data dma_stm32_data_##index section

**Review by erwango — APPROVED** [review]

**Pull request merged**